### PR TITLE
Add minimum Fee Bump replacement fee to `POST /tx`'s  response

### LINKED ISSUE #2505: Add Fee-Recommendation to Transaction Submission Response
After the upgrade to protocol 13, it will be possible to submit fee-bump transactions. Fee-bump transactions will participate in replace-by-fee semantics with back-off. It is possible, as a consequence, to submit a fee-bump transaction to stellar-core and have it rejected as invalid due to an insufficiently large ratio between its fee and the fee of a transaction in the queue. It is currently difficult (perhaps impossible) to know what fee to submit if this happens, so we should add information about this in the response.

### PR DESCRIPTION
**Pull request by 2opremio**
# Description

Fixes #2505

When the fee of a Fee Bump transaction isn't high enough to satisfy the replace-by-fee requirement, include ~a fee-rate hint to~ the minimum Fee Bump replacement fee to:

1. Be more precise on why the fee wasn't enough.
2. Give the user an opportunity to recreate a transaction with an acceptable fee (although the fee requirement is dynamic, so, there are no guarantees).

The ~hint~ minimum fee is represented as an ~rational number (i.e. providing a numerator and denominator)~  integer in the JSON response. Here is an example:

```
"min_fee": 11,
```

This approach is reactive (i.e. the user needs to try submitting a transaction in order get ~a hint~ the minimum).

Alternatively we could add a separate endpoint (or a query parameter) to only check the fee required. I think being reactive is sufficient and adding a preliminary check adds unnecessary complexity, but, I am happy to be proven wrong.

I am really looking forward to your feedback. This is my first Core PR and I haven't done C++ in 12 years, so be prepared to read cringy code.

Tests will be added once a get a GA on the approach.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Comment by jonjove** [comment]
Thanks for submitting this! I don't really see any big C++ issues here @2opremio. However, I do have a few comments/questions about the PR:

1. It doesn't build on my local machine (errors in HerderTests.cpp)
2. I think we should probably add a new `AddResult` enumeration for this bad replace-by-fee case with a name like `ADD_STATUS_BAD_REPLACE_BY_FEE` or `ADD_STATUS_REPLACE_BY_FEE_INSUFFICIENT`. The HTTP endpoint shouldn't return the `TransactionResult` in this case, it should just return the fee recommendation
3. I think you can mitigate a lot of collateral damage by having two overloads of `Herder::recvTransaction`: the one that already exists, and one that uses pass-by-reference to return the `feeRec`. Most existing code doesn't care about the `feeRec` and so will continue to use the original overload, avoiding changes in those places. The code that does care (eg. in `CommandHandler::tx`) will use the new overload with `feeRec`. In general, using pass-by-reference in this PR might simplify a lot of things.
4. What advantage do you see in returning the recommended fee-per-operation (as a rational) instead of the recommended fee (as an integer)? I imagine that in basically every single case you will want to know the recommended fee for the transaction that you submitted, rather than using it to calculate the recommended fee for some other transaction. In fact, you generally don't want to use this for other transactions because they will only have the same fee recommendation if they have the same source account and sequence number. But maybe I'm missing something here?

**Comment by 2opremio** [comment]
Thanks for the review!

> 1. It doesn't build on my local machine (errors in HerderTests.cpp)

I fixed that. But now I am getting a segmentation fault in one of the tests:

```
1	-------------------------------------------------------------------------------
1	LedgerCloseMetaStream file descriptor - REPLAY_IN_MEMORY
1	-------------------------------------------------------------------------------
1	ledger/test/LedgerCloseMetaStreamTests.cpp:105
1	...............................................................................
1	
1	ledger/test/LedgerCloseMetaStreamTests.cpp:105: FAILED:
1	due to a fatal error condition:
1	  SIGSEGV - Segmentation violation signal
1	
1	<done>
```

I doubt my new could has caused incorrect memory accesses. Any suggestions on how to debug this?

> 2. I think we should probably add a new `AddResult` enumeration for this bad replace-by-fee case with a name like `ADD_STATUS_BAD_REPLACE_BY_FEE` or `ADD_STATUS_REPLACE_BY_FEE_INSUFFICIENT`. The HTTP endpoint shouldn't return the `TransactionResult` in this case, it should just return the fee recommendation

Done

> 3. I think you can mitigate a lot of collateral damage by having two overloads of `Herder::recvTransaction`: the one that already exists, and one that uses pass-by-reference to return the `feeRec`. Most existing code doesn't care about the `feeRec` and so will continue to use the original overload, avoiding changes in those places. The code that does care (eg. in `CommandHandler::tx`) will use the new overload with `feeRec`. In general, using pass-by-reference in this PR might simplify a lot of things.

I've done this, but I am not sure it's an improvement (the PR has become even larger).

> 4. What advantage do you see in returning the recommended fee-per-operation (as a rational) instead of the recommended fee (as an integer)? I imagine that in basically every single case you will want to know the recommended fee for the transaction that you submitted, rather than using it to calculate the recommended fee for some other transaction. In fact, you generally don't want to use this for other transactions because they will only have the same fee recommendation if they have the same source account and sequence number. But maybe I'm missing something here?

After thinking about it, there is no advantage. I changed the code.

**Comment by MonsieurNicolas** [comment]
great.

wrt what we return:
* agree with Jon, there is no need to expose fee rate or some new notation for fees. If what we return is the smallest fee amount that would allow the fee bump to be accepted, we can just use an amount.
   *  (I see that you fixed that already)
* (minor) we should not call this a recommendation internally or in the contract exposed to the outside, it's a statement about the current state. I'd suggest using terminology like `feeBumpAcceptableFee` or `feeBumpMinFee`. A recommendation could be derived from this at the SDK level by applying an extra multiplier on top but this is not a core concern.

File: src/herder/TransactionQueue.h [hunk 1]
```diff
@@ -59,13 +59,20 @@ class TransactionQueue
...
-    enum class AddResult
+    enum class AddStatus
     {
         ADD_STATUS_PENDING = 0,
         ADD_STATUS_DUPLICATE,
         ADD_STATUS_ERROR,
         ADD_STATUS_TRY_AGAIN_LATER,
-        ADD_STATUS_COUNT
+        ADD_STATUS_COUNT,
+        ADD_STATUS_BAD_REPLACE_BY_FEE,
```
Comment (MonsieurNicolas) [inline src/herder/TransactionQueue.h:69]: this seems wrong: `ADD_STATUS_COUNT` is used as the marker for the upper bound for the enum values

Comment (2opremio) [reply src/herder/TransactionQueue.h:68]: Good catch, thanks. It should be fixed now

File: src/simulation/LoadGenerator.cpp [hunk 2]
```diff
@@ -630,14 +630,14 @@ LoadGenerator::TxInfo::execute(Application& app, bool isCreate,
     msg.transaction() = txf->getEnvelope();
     txm.mTxnBytes.Mark(xdr::xdr_argpack_size(msg));
 
-    auto status = app.getHerder().recvTransaction(txf);
-    if (status != TransactionQueue::AddResult::ADD_STATUS_PENDING)
+    auto result = app.getHerder().recvTransaction(txf);
```
Comment (jonjove) [inline src/simulation/LoadGenerator.cpp:633]: Nitpick: why did you rename this from `status` to `result`? It is antagonistic to your other change of renaming `AddResult` to `AddStatus`.

File: src/overlay/Peer.cpp [hunk 3]
```diff
@@ -784,15 +784,16 @@ Peer::recvTransaction(StellarMessage const& msg)
     {
         // add it to our current set
         // and make sure it is valid
-        auto recvRes = mApp.getHerder().recvTransaction(transaction);
+        auto recvResStatus = mApp.getHerder().recvTransaction(transaction);
```
Comment (jonjove) [inline src/overlay/Peer.cpp:787]: Nitpick: The name reads as "receive result status" which is weird. How about just `recvStatus`?

File: src/main/CommandHandler.cpp [hunk 4]
```diff
@@ -568,22 +568,28 @@ CommandHandler::tx(std::string const& params, std::string& retStr)
...
             }
-
-            output << "{"
-                   << "\"status\": "
-                   << "\"" << TX_STATUS_STRING[static_cast<int>(status)]
-                   << "\"";
-            if (status == TransactionQueue::AddResult::ADD_STATUS_ERROR)
+            break;
+            case TransactionQueue::AddStatus::ADD_STATUS_BAD_REPLACE_BY_FEE:
+                output << " , \"min_fee\": " << feeBumpMinFee;
```
Comment (jonjove) [inline src/main/CommandHandler.cpp:590]: Do we need to use a string here? Otherwise, will this work correctly with languages that don't support `int64` natively?

File: src/herder/TransactionQueue.cpp [hunk 5]
```diff
@@ -60,8 +60,19 @@ canReplaceByFee(TransactionFrameBasePtr tx, TransactionFrameBasePtr oldTx)
...
     auto v1 = bigMultiply(newFee, oldNumOps);
-    auto v2 = bigMultiply(oldFee, newNumOps);
-    return v1 >= TransactionQueue::FEE_MULTIPLIER * v2;
+    auto v2 = uint128_t(TransactionQueue::FEE_MULTIPLIER) *
+              bigMultiply(oldFee, newNumOps);
+    std::pair<bool, int64_t> result;
+    result.first = v1 >= v2;
+    if (!result.first)
+    {
+        // Add a fee recommendation for future attempts
```
Comment (jonjove) [inline src/herder/TransactionQueue.cpp:69]: As @MonsieurNicolas pointed out, we shouldn't use the term "fee recommendation" anywhere.

File: src/herder/TransactionQueue.cpp [hunk 6]
```diff
@@ -60,8 +60,19 @@ canReplaceByFee(TransactionFrameBasePtr tx, TransactionFrameBasePtr oldTx)
...
+        // recommendation = FEE_MULTIPLIER * oldFeeRate * newNumOps
+        //                = FEE_MULTIPLIER * (oldFee / oldNumOps) * newNumOps
+        //                = (FEE_MULTIPLIER * oldFee * newNumOps) / oldNumOps
+        result.second = bigDivide(v2, oldNumOps, ROUND_UP);
```
Comment (MonsieurNicolas) [inline src/herder/TransactionQueue.cpp:73]: Can we maybe reorganize the code a bit differently:
only one code path cares about this computation, so we could move it to a function `computeMinFee` of its own that gets called by the caller at the top if it runs into the `txINSUFFICIENT_FEE` error (see my other commend in the `tx` endpoint, related).

Reason this came up during my review is that `bigDivide` can throw and I'd hate to crash the other code paths.

File: src/main/CommandHandler.cpp [hunk 7]
```diff
@@ -568,22 +568,28 @@ CommandHandler::tx(std::string const& params, std::string& retStr)
...
                 mApp.getOverlayManager().broadcastMessage(msg);
             }
-
-            output << "{"
-                   << "\"status\": "
-                   << "\"" << TX_STATUS_STRING[static_cast<int>(status)]
-                   << "\"";
-            if (status == TransactionQueue::AddResult::ADD_STATUS_ERROR)
+            break;
+            case TransactionQueue::AddStatus::ADD_STATUS_BAD_REPLACE_BY_FEE:
```
Comment (MonsieurNicolas) [inline src/main/CommandHandler.cpp:589]: Sorry, I should have noticed this much earlier. I just realized something that impacts the overall structure of this PR.

I don't think we need to add a new internal status result... or even a `minFee` element in the json:

what we really want here is that if the transaction fails with `txINSUFFICIENT_FEE`, we provide the "mininum fee" required for that transaction.

For replace by fee, it's the computation that you've done, for others, it depends on context (grep for `tx`).

OK, so if we want to always return a min fee with this kind of error... how do we return it?

For that, I suspect we can simply use `feeCharged` (in the result) to propagate the min fee on error from anywhere in the code (NB: it may already be the case for the other failures), instead of having special handling for fee bump failures coming from the mempool.

It's probably safe to do as fees are always charged before applying transactions (true now and for all older protocol versions).

Having this included in the result is actually quite nice if we can do it because I imagine that we already propagate back the raw result to the SDKs so the changes to Horizon & clients should be minimal.

File: src/main/CommandHandler.cpp (hunk 4 above)
Comment (MonsieurNicolas) [reply src/main/CommandHandler.cpp:590]: +1: this is unsafe

File: src/main/CommandHandler.cpp (hunk 7 above)
Comment (jonjove) [reply src/main/CommandHandler.cpp:589]: I don’t really think this is a good idea. First, I think it would be unnecessary. If the fee is insufficient for any reason other than replace-by-fee backoff then you can just calculate the required fee using the base fee from the ledger header. In that case, stellar-core doesn’t need to give you any new information. 

Second, I think there is already a huge amount of confusion regarding the results of transaction submission and validation. Your proposal definitely wouldn’t help reduce that confusion, and it could make it worse. For example, people might wonder why stellar-core is charging a fee higher than the fee specified in the transaction (it isn’t, of course, it’s just telling you what fee would have been required). This is just a single way this could be misinterpreted. But good interfaces should be easy to use correctly and hard to use incorrectly. The interface as described in @2opremio PR is obvious and unambiguous:

- You only get that result if you submit a fee-bump (this was part of your motivation of making replace-by-fee only happen for fee-bumps) so there are no backwards compatibility issues
- The result clearly indicates what happened and cannot be misinterpreted

If you really wanted to always provide fee information whenever a transaction is rejected because the fee is bad, I would actually argue that we should add `ADD_STATUS_BAD_FEE` instead of `ADD_STATUS_BAD_REPLACE_BY_FEE` and provide information there. But again, I think this is unnecessary.

Comment (MonsieurNicolas) [reply src/main/CommandHandler.cpp:589]: There are a few things here:
* the main point is that this proposal is "replace by fee" centric, and is not future proof.
* on the contract front
  * we are already returning a transaction result, and `feeCharged` is already set to some arbitrary thing. If we were to actually formally define what the value of `feeCharged` is on failure, I think it would be a net improvement over the current situation where people don't realize that it's the same.
  * that said, if we were to add a new field on failure because we're not happy with transaction result, we should just make the interface harder to use wrong (and it would not have two "fees" returned in one call). I had a proposal to remove using transaction results altogether for this kind of failure but there was not much appetite for having to support a different contract from the Horizon team (I think because it involves changing SDKs). This change requires changing SDKs etc - so maybe it's the right time.

Comment (jonjove) [reply src/main/CommandHandler.cpp:589]: @2opremio After looking at the transaction result contract more carefully, I think the approach mentioned by @MonsieurNicolas is a good idea. It'll be a lot less code, and downstream systems will already be wired up for it. Long run, I still think we should be making this interface less confusing as noted in #1999. Sorry about the change of plans.

**Comment by 2opremio** [comment]
closing in favor of https://github.com/stellar/stellar-core/pull/2624
